### ISSUE TITLE
Don't emit invalid-name for classes inside functions

### ISSUE THREAD

**Issue by The-Compiler** [PyCQA/pylint#1049]
Sometimes, it's useful to define classes inside a function - for example like this:

``` python
def _generate_cmdline_tests():
    TestCase = collections.namedtuple('TestCase', 'cmd, valid')
    separators = [';;', ' ;; ', ';; ', ' ;;']
    invalid = ['foo', '']
    valid = ['leave-mode', 'hint all']
    # Valid command only -> valid
    for item in valid:
        yield TestCase(''.join(item), True)
    # [...]
```

or to have a local alias for something:

``` python
def test_tokenize_x11_geometry_string_manual():
    Token = utils.X11GeometryStringParser.Token
    parser = utils.X11GeometryStringParser()
    res = list(parser.tokenize("80x24+30+200"))
    assert res == [
        Token(typ='NUMBER', value='80', column=0),
        Token(typ='MUL', value='x', column=2),
        # [...]
    ]
```

However, pylint treats those as variables and emits `invalid-name` if they don't satisfy `variable-rgx`.

cc @ganwell

**Comment by PCManticore** [comment]
Makes sense.

**Comment by AWhetter** [comment]
To any new contributors, this will require adding a clause to the base checker (https://github.com/PyCQA/pylint/blob/fbdf5ba095ebbbf04db12f6c3e88e21152a30507/pylint/checkers/base.py#L1279) where we check if the assignment value looks like a class. This can be done by inferring it and checking if it is a ClassDef. If it is uninferable you could maybe fall back to checking if the name is "class-like".
